packages/@aws-cdk/aws-codebuild/lib/source.ts

```diff
@@ -6,73 +6,150 @@ import { cloudformation } from './codebuild.generated';
+    return {
+      sourceIdentifier: this.identifier,
+      type: this.type,
+      ...sourceProp,
+    };
+  }
+
+  protected toSourceProperty(): any {
+    return {
+    };
+  }
 }
 
+/**
+ * A `NO_SOURCE` CodeBuild Project Source definition.
+ * This is the default source type,
+ * if none was specified when creating the Project.
+ * *Note*: the `NO_SOURCE` type cannot be used as a secondary source,
+ * and because of that, you're not allowed to specify an identifier for it.
+ */
 export class NoSource extends BuildSource {
+  public readonly type: SourceType = SourceType.None;
+
   constructor() {
-    super();
+    super({});
   }
+}
 
-  public toSourceJSON(): cloudformation.ProjectResource.SourceProperty {
-    return {
-      type: SourceType.None,
-    };
-  }
+/**
+ * Construction properties for {@link CodeCommitSource}.
+ */
+export interface CodeCommitSourceProps extends BuildSourceProps {
+  repository: codecommit.RepositoryRef;
 }
 
 /**
- * CodeCommit Source definition for a CodeBuild project
+ * CodeCommit Source definition for a CodeBuild project.
  */
 export class CodeCommitSource extends BuildSource {
-  constructor(private readonly repo: codecommit.RepositoryRef) {
-    super();
+  public readonly type: SourceType = SourceType.CodeCommit;
+  private readonly repo: codecommit.RepositoryRef;
+
+  constructor(props: CodeCommitSourceProps) {
+    super(props);
+    this.repo = props.repository;
   }
 
-  public bind(project: Project) {
+  public _bind(project: Project) {
     // https://docs.aws.amazon.com/codebuild/latest/userguide/setting-up.html
     project.addToRolePolicy(new iam.PolicyStatement()
       .addAction('codecommit:GitPull')
       .addResource(this.repo.repositoryArn));
   }
 
-  public toSourceJSON(): cloudformation.ProjectResource.SourceProperty {
+  protected toSourceProperty(): any {
     return {
-      type: SourceType.CodeCommit,
       location: this.repo.repositoryCloneUrlHttp
     };
   }
 }
 
 /**
- * CodePipeline Source definition for a CodeBuild project
+ * Construction properties for {@link S3BucketSource}.
  */
-export class CodePipelineSource extends BuildSource {
-  public toSourceJSON(): cloudformation.ProjectResource.SourceProperty {
+export interface S3BucketSourceProps extends BuildSourceProps {
+  bucket: s3.BucketRef;
+  path: string;
+}
+
+/**
+ * S3 bucket definition for a CodeBuild project.
+ */
+export class S3BucketSource extends BuildSource {
+  public readonly type: SourceType = SourceType.S3;
+  private readonly bucket: s3.BucketRef;
+  private readonly path: string;
+
+  constructor(props: S3BucketSourceProps) {
+    super(props);
+    this.bucket = props.bucket;
+    this.path = props.path;
+  }
+
+  public _bind(project: Project) {
+    this.bucket.grantRead(project.role);
+  }
+
+  protected toSourceProperty(): any {
     return {
-      type: SourceType.CodePipeline
+      location: new cdk.FnConcat(this.bucket.bucketName, '/', this.path)
```

Comment:
We can do TypeScript string interpolation these days and the world is great:

```
location: `${this.bucket.bucketName}/${this.path}`
```

Reply:
Amazing 😱